src/streaming-feed.ts

```diff
@@ -0,0 +1,149 @@
+import { BatchId, Bee, Reference, Signer, Topic, Utils } from '@ethersphere/bee-js'
+import { FeedType } from './feed'
+import {
+  assembleSocPayload,
+  makeTopic,
+  mapSocToFeed,
+  StreamingFeedChunk,
+  SwarmStreamingFeed,
+  SwarmStreamingFeedR,
+  SwarmStreamingFeedRW,
+} from './streaming'
+import { ChunkReference, makeSigner, writeUint64BigEndian } from './utils'
+
+const { Hex } = Utils
+const { hexToBytes } = Hex
+export const getCurrentTime = (d = new Date()) => d.getTime()
+export class StreamingFeed implements SwarmStreamingFeed<number> {
+  public readonly type: FeedType
+
+  public constructor(public readonly bee: Bee) {
+    this.type = 'fault-tolerant-stream'
+  }
+
+  public makeFeedR(
+    topic: Topic | Uint8Array | string,
+    owner: Utils.Eth.EthAddress | Uint8Array | string,
+  ): SwarmStreamingFeedR<number> {
+    const socReader = this.bee.makeSOCReader(owner)
+    const topicHex = makeTopic(topic)
+    const topicBytes = hexToBytes<32>(topicHex)
+    const ownerHex = Utils.Eth.makeHexEthAddress(owner)
+
+    const getIndexForArbitraryTime = async (
+      lookupTime: number,
+      initialTime: number,
+      updatePeriod: number,
+    ): Promise<number> => {
+      const currentTime = getCurrentTime() // Tp
+      try {
+        //  the nearest last index to an arbitrary time (Tx) where T0 <= Tx <= Tn <= Tp
+        if (currentTime >= initialTime && lookupTime >= initialTime) {
+          return Math.floor((lookupTime - initialTime) / updatePeriod)
+        }
+      } catch (e) {
+        // no-op
+      }
+
+      return -1
+    }
+
+    // Download Feed Chunk at Specific Time
+    const getUpdate = async (
+      initialTime: number,
+      updatePeriod: number,
+      lookupTime?: number,
+    ): Promise<StreamingFeedChunk> => {
+      lookupTime = lookupTime ?? getCurrentTime()
+      const index = await getIndexForArbitraryTime(lookupTime, initialTime, updatePeriod)
+      const socChunk = await socReader.download(this.getIdentifier(topicBytes, index))
+
+      return mapSocToFeed(socChunk)
+    }
+
+    //  Download Feed Stream
+    const getUpdates = async (initialTime: number, updatePeriod: number): Promise<StreamingFeedChunk[]> => {
+      const feeds: StreamingFeedChunk[] = []
+      // while from last to first, use lookupTime = chunk.timestamp + 1
+
+      const index = await getIndexForArbitraryTime(getCurrentTime(), initialTime, updatePeriod)
+      const socChunk = await socReader.download(this.getIdentifier(topicBytes, index - 1))
+
+      let feed = await mapSocToFeed(socChunk)
+      let i = 1
+      feeds.push(feed)
+      while (i < index) {
+        feed = await getUpdate(initialTime, updatePeriod, feed.timestamp + updatePeriod + 1)
```

Comment:
I cannot comprehend with this section here.
you downloaded the penultimate chunk from which you go _forward_ in updates that you repeat times based on the index of the penultimate chunk...
can you elaborate please? maybe I'm missing something.